File: api/src/functions/emailShoppingList.ts
import { app, HttpRequest, HttpResponseInit, InvocationContext } from '@azure/functions'
import { getById } from '../cosmosdb.js'
import { getTroopContext, unauthorized, forbidden } from '../middleware/auth.js'
import { checkPermission } from '../middleware/roles.js'
import { emailShoppingListSchema, validationError } from '../schemas.js'

const EVENTS_CONTAINER = 'events'
const SENDGRID_API_URL = 'https://api.sendgrid.com/v3/mail/send'

function escapeHtml(value: string): string {
  return value
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;')
}

async function emailShoppingListHandler(req: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
  const eventId = req.params.id
  context.log(`POST /api/events/${eventId}/shopping-list/email`)

  const auth = await getTroopContext(req, context)
  if (!auth) return unauthorized()
  if (!checkPermission(auth.role, 'viewContent')) return forbidden()

  try {
    const parsed = emailShoppingListSchema.safeParse(await req.json())
    if (!parsed.success) return validationError(parsed.error)

    const existingEvent = await getById<any>(EVENTS_CONTAINER, eventId, auth.troopId)
    if (!existingEvent) return { status: 404, jsonBody: { error: 'Event not found' } }

    const sendGridApiKey = process.env.SENDGRID_API_KEY
    const fromEmail = process.env.SENDGRID_FROM_EMAIL

    if (!sendGridApiKey || !fromEmail) {
      context.error('Missing SENDGRID_API_KEY or SENDGRID_FROM_EMAIL')
      return { status: 500, jsonBody: { error: 'Email service not configured' } }
    }

    const { recipientEmail, items } = parsed.data
    const resolvedEventName = typeof existingEvent.name === 'string' && existingEvent.name.length > 0
      ? existingEvent.name
      : 'Unnamed Event'
    const textRows = items.map((item) => `- ${item.name}: ${item.quantity} ${item.unit}`).join('\n')
    const htmlRows = items
      .map((item) => {
        const safeName = escapeHtml(item.name)
        const safeQuantity = escapeHtml(String(item.quantity))
        const safeUnit = escapeHtml(item.unit)
        return `<tr><td style="padding:4px 8px;">${safeName}</td><td style="padding:4px 8px;text-align:right;">${safeQuantity}</td><td style="padding:4px 8px;">${safeUnit}</td></tr>`
      })
      .join('')

    const response = await fetch(SENDGRID_API_URL, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${sendGridApiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        personalizations: [{ to: [{ email: recipientEmail }] }],
        from: { email: fromEmail },
        subject: `Shopping List: ${resolvedEventName}`,
        content: [
          {
            type: 'text/plain',
            value: `Shopping list for ${resolvedEventName}\n\n${textRows}`,
          },
          {
            type: 'text/html',
            value: `<h2>Shopping list for ${escapeHtml(resolvedEventName)}</h2><table style="border-collapse:collapse;"><thead><tr><th style="padding:4px 8px;text-align:left;">Item</th><th style="padding:4px 8px;text-align:right;">Quantity</th><th style="padding:4px 8px;text-align:left;">Unit</th></tr></thead><tbody>${htmlRows}</tbody></table>`,
          },
        ],
      }),
    })

    if (!response.ok) {
      const sendGridError = await response.text()
      context.error('SendGrid request failed:', sendGridError)
      return { status: 502, jsonBody: { error: 'Failed to send email' } }
    }

    return { status: 202, jsonBody: { message: 'Shopping list email sent' } }
  } catch (err) {
    context.error(`POST /api/events/${eventId}/shopping-list/email failed:`, err)
    const message = err instanceof Error ? err.message : String(err)
    return { status: 500, jsonBody: { error: message } }
  }
}

app.http('emailShoppingList', {
  methods: ['POST'],
  authLevel: 'anonymous',
  route: 'events/{id}/shopping-list/email',
  handler: emailShoppingListHandler,
})

File: api/app/routers/email_shopping_list.py
from __future__ import annotations

import html
import logging
import os

from azure.communication.email import EmailClient
from azure.identity import DefaultAzureCredential
from fastapi import APIRouter, HTTPException

from app.cosmosdb import get_by_id
from app.middleware.auth import RequireTroopContext, forbidden
from app.middleware.roles import check_permission
from app.schemas import EmailShoppingList

logger = logging.getLogger(__name__)
router = APIRouter()

EVENTS_CONTAINER = "events"
MAX_SUBJECT_EVENT_NAME_LENGTH = 200

_email_client: EmailClient | None = None


def _get_email_client() -> EmailClient:
    global _email_client
    if _email_client is None:
        endpoint = os.environ.get("ACS_ENDPOINT")
        if not endpoint:
            raise RuntimeError("ACS_ENDPOINT is not configured")
        _email_client = EmailClient(f"https://{endpoint}", DefaultAzureCredential())
    return _email_client


@router.post("/events/{event_id}/shopping-list/email", status_code=202)
async def email_shopping_list(event_id: str, body: EmailShoppingList, auth: RequireTroopContext):
    if not check_permission(auth.role, "viewContent"):
        forbidden()

    existing_event = await get_by_id(EVENTS_CONTAINER, event_id, auth.troopId)
    if not existing_event:
        raise HTTPException(status_code=404, detail="Event not found")

    from_email = os.environ.get("ACS_FROM_EMAIL")
    if not from_email:
        raise HTTPException(status_code=500, detail="Email service not configured")

    try:
        client = _get_email_client()
    except RuntimeError:
        raise HTTPException(status_code=500, detail="Email service not configured")

    event_name = existing_event.get("name") or "Unnamed Event"
    safe_event_name = str(event_name).replace("\n", " ").replace("\r", " ")[:MAX_SUBJECT_EVENT_NAME_LENGTH]

    text_rows = "\n".join(f"- {item.name}: {item.quantity} {item.unit}" for item in body.items)
    html_rows = "".join(
        f'<tr><td style="padding:4px 8px;">{html.escape(item.name)}</td>'
        f'<td style="padding:4px 8px;text-align:right;">{html.escape(str(item.quantity))}</td>'
        f'<td style="padding:4px 8px;">{html.escape(item.unit)}</td></tr>'
        for item in body.items
    )

    message = {
        "senderAddress": from_email,
        "recipients": {"to": [{"address": body.recipientEmail}]},
        "content": {
            "subject": f"Shopping List: {safe_event_name}",
            "plainText": f"Shopping list for {safe_event_name}\n\n{text_rows}",
            "html": (
                f"<h2>Shopping list for {html.escape(safe_event_name)}</h2>"
                f'<table style="border-collapse:collapse;">'
                f'<thead><tr><th style="padding:4px 8px;text-align:left;">Item</th>'
                f'<th style="padding:4px 8px;text-align:right;">Quantity</th>'
                f'<th style="padding:4px 8px;text-align:left;">Unit</th></tr></thead>'
                f"<tbody>{html_rows}</tbody></table>"
            ),
        },
    }

    poller = client.begin_send(message)
    result = poller.result()

    if result.get("status") != "Succeeded":
        raise HTTPException(status_code=502, detail="Failed to send email")

    return {"message": "Shopping list email sent"}
